server/src/main/java/org/opensearch/index/store/remote/utils/cache/SegmentedCache.java

```diff
@@ -0,0 +1,338 @@
+import org.opensearch.common.cache.Weigher;
+import org.opensearch.index.store.remote.utils.cache.stats.CacheStats;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiFunction;
+
+/**
+ * Segmented {@link LRUCache} to offer concurrent access with less contention.
+ * @param <K> type of the key
+ * @param <V> type of th value
+ *
+ * @opensearch.internal
+ */
+public class SegmentedCache<K, V> implements RefCountedCache<K, V> {
+
+    private static final int HASH_BITS = 0x7fffffff; // usable bits of normal ActivenessAwareCache hash
+
+    private static final int ceilingNextPowerOfTwo(int x) {
+        // From Hacker's Delight, Chapter 3, Harry S. Warren Jr.
+        return 1 << (Integer.SIZE - Integer.numberOfLeadingZeros(x - 1));
+    }
+
+    private final long capacity;
+
+    private final long perSegmentCapacity;
+    /**
+     * The array of bins. Size is always a power of two.
+     */
+    private final RefCountedCache<K, V>[] table;
+
+    /**
+     * Mask value for indexing into segments.
+     */
+    private final int segmentMask;
+
+    private final Weigher<V> weigher;
+
+    public SegmentedCache(Builder<K, V> builder) {
+        this.capacity = builder.capacity;
+        final int segments = ceilingNextPowerOfTwo(builder.concurrencyLevel);
+        this.segmentMask = segments - 1;
+        this.table = newSegmentArray(segments);
+        this.perSegmentCapacity = (capacity + (segments - 1)) / segments;
+        this.weigher = builder.weigher;
+        for (int i = 0; i < table.length; i++) {
+            table[i] = new LRUCache<>(perSegmentCapacity, builder.listener, builder.weigher);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    final RefCountedCache<K, V>[] newSegmentArray(int size) {
+        return new RefCountedCache[size];
+    }
+
+    RefCountedCache<K, V> segmentFor(K key) {
+        int h = key.hashCode();
+        // Spreads (XORs) higher bits of hash to lower and also forces top
+        // bit to 0. Because the table uses power-of-two masking, sets of
+        // hashes that vary only in bits above the current mask will
+        // always collide. (Among known examples are sets of Float keys
+        // holding consecutive whole numbers in small tables.) So we
+        // apply a transform that spreads the impact of higher bits
+        // downward. There is a tradeoff between speed, utility, and
+        // quality of bit-spreading. Because many common sets of hashes
+        // are already reasonably distributed (so don't benefit from
+        // spreading), and because we use trees to handle large sets of
+        // collisions in bins, we just XOR some shifted bits in the
+        // cheapest possible way to reduce systematic lossage, as well as
+        // to incorporate impact of the highest bits that would otherwise
+        // never be used in index calculations because of table bounds.
+        return table[(h ^ (h >>> 16)) & HASH_BITS & segmentMask];
+    }
```

Comment:
`ConcurrentHashMap` does this since it degrades to a red-black tree if there are too many collisions. It is not a good distribution and should not be used elsewhere. Thomas Mueller's hash function (of H2 fame) is a good default.

```java
int spread(int x) {
  x = ((x >>> 16) ^ x) * 0x45d9f3b;
  x = ((x >>> 16) ^ x) * 0x45d9f3b;
  return (x >>> 16) ^ x;
}
```

Reply:
applied your suggestion in next commit. Please review it again since I'm not an expert here

Reply:
lgtm. thanks!

Thomas uses simulated annealing to compute that hash, which was rediscovered in [Prospecting for Hash Functions](https://nullprogram.com/blog/2018/07/31/) with some other good variations. I used Thomas' in Caffeine's frequency histogram (4-bit CountMinSketch) and recently switched to prospector's 3 round hash when rewriting it to optimize for the cpu cache line (to reduce random memory access lookups).